README.md
# NAME

Device::LSM303DLHC - I2C interface to LSM303DLHC 3 axis magnetometer(compass) and accelerometer using Device::SMBus

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-brightgreen.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/shantanubhadoria/perl-Device-LSM303DLHC"><img src="https://api.travis-ci.org/shantanubhadoria/perl-Device-LSM303DLHC.svg?branch=build/master" alt="Travis status" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Device-LSM303DLHC%200.014"><img src="https://badgedepot.code301.com/badge/cpantesters/Device-LSM303DLHC/0.014" alt="CPAN Testers result" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Device-LSM303DLHC-0.014"><img src="https://badgedepot.code301.com/badge/kwalitee/Device-LSM303DLHC/0.014" alt="Distribution kwalitee" /></a>
    <a href="https://gratipay.com/shantanubhadoria"><img src="https://img.shields.io/gratipay/shantanubhadoria.svg" alt="Gratipay" /></a>
    </p>
</div>

# VERSION

version 0.014

# ATTRIBUTES

## I2CBusDevicePath

this is the device file path for your I2CBus that the LSM303DLHC is connected on e.g. /dev/i2c-1
This must be provided during object creation.

## Magnetometer

    $self->Magnetometer->enable();
    $self->Magnetometer->getReading();

This is a object of [Device::Magnetometer::LSM303DLHC](https://metacpan.org/pod/Device::Magnetometer::LSM303DLHC)

## Accelerometer 

    $self->Accelerometer->enable();
    $self->Accelerometer->getReading();

This is a object of [Device::Accelerometer::LSM303DLHC](https://metacpan.org/pod/Device::Accelerometer::LSM303DLHC)

# SUPPORT

## Bugs / Feature Requests

Please report any bugs or feature requests through github at 
[https://github.com/shantanubhadoria/perl-device-lsm303dlhc/issues](https://github.com/shantanubhadoria/perl-device-lsm303dlhc/issues).
You will be notified automatically of any progress on your issue.

## Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

[https://github.com/shantanubhadoria/perl-device-lsm303dlhc](https://github.com/shantanubhadoria/perl-device-lsm303dlhc)

    git clone git://github.com/shantanubhadoria/perl-device-lsm303dlhc.git

# AUTHOR

Shantanu Bhadoria &lt;shantanu at cpan dott org>

# CONTRIBUTOR

Shantanu <shantanu@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Changelog for Device-LSM303DLHC

0.014 2016-04-07T06:45:58
 - No changes found

0.013 2015-03-25T07:20:05
 - v0.013 (Shantanu Bhadoria)
 - Reorganising, extending instead of using (Shantanu Bhadoria)

0.012 2015-03-25T07:18:39
 - v0.012 (Shantanu Bhadoria)
 - Reorganising, extending instead of using (Shantanu Bhadoria)
 - Fixed Raw reading on scale 1 (Shantanu Bhadoria)

0.011 2013-10-05T01:27:58
 - v0.011 (Shantanu)
 - small changes (Shantanu Bhadoria)

0.010 2013-10-03T15:24:10
 - v0.010 (Shantanu)
 - Non Raw reading scaled between -0.5 and 0.5 (Shantanu)

0.009 2013-10-03T15:22:09
 - v0.009 (Shantanu)
 - Non Raw reading scaled between -0.5 and 0.5 (Shantanu)

0.008 2013-09-29T02:48:36
 - v0.008 (Shantanu)
 - POD spelling fixes (Shantanu)

0.007 2013-09-27T11:41:12
 - v0.007 (Shantanu)
 - Renamed Compass to Magnetometer to use the more scientific term since
   the term compass is usually used for 2D magnetometers not 3D
   Magnetometers. (Shantanu)

0.006 2013-09-27T10:50:09
 - v0.006 (Shantanu)
 - Removed Math::Trig usage. manually computing trig functions (Shantanu)
 - Resolved xt test failures (Shantanu)
 - Refactoring Submodule to allow easy loading of similar subdevices as
   plugins using a common prefix in MOP (Shantanu Bhadoria)
 - minor changes (Shantanu Bhadoria)

0.005 2013-09-12T09:23:58
 - v0.005 (Shantanu)
 - Added README.pod and Changes (Shantanu)

0.004 2013-09-12T06:47:10
 - v0.004 (Shantanu)
 - Moved to ChangelogFromGit compliance with CPAN::Changes::Spec (Shantanu)

0.003 2013-09-12T06:11:56
 - Moved to ChangelogFromGit compliance with CPAN::Changes::Spec (Shantanu)

0.002 2013-09-11T09:15:31
 - pod fix (Shantanu)

0.001 2013-09-11T07:46:23
 -  final mods to the pod` (Shantanu)
 - Fixed bugs in core module (Shantanu)
 -  minor commit (Shantanu Bhadoria)
 - First Commit (Shantanu Bhadoria)
 - First Push with basic Raw Readings from LSM303DLHC Compass and
   Accelerometer (Shantanu Bhadoria)


examples/example.pl
use strict;
use warnings;
use FindBin qw($Bin);
use lib "$Bin/../lib";

use Device::LSM303DLHC;
#use Device::Compass::LSM303DLHC;
#use Device::Accelerometer::LSM303DLHC;

my $dev = Device::LSM303DLHC->new(I2CBusDevicePath => '/dev/i2c-1');
$dev->Compass->enable();
$dev->Accelerometer->enable();
use Data::Dumper;
my ($minx, $miny, $minz, $maxx, $maxy, $maxz) = (500,500,500,0,0,0);
my ($avg,$avgCount) = (0,0);
while(){
    my $accelerometer = $dev->Accelerometer->getRawReading();
    #my $accelerometer = $dev->Accelerometer->getAccelerationVectorInG();
    #my $accelerometer = $dev->Accelerometer->getAccelerationVectorInMSS();
    my $compass = $dev->Compass->getRawReading();
    my $accAngle = $dev->Accelerometer->getAccelerationVectorAngles();


    #print "COMPASS: $compass->{x}\t$compass->{y}\t$compass->{z}\t\n";
    print "\tACCELEROMETER: $accelerometer->{x}\t$accelerometer->{y}\t$accelerometer->{z}\t\tCOMPASS: $compass->{x}\t$compass->{y}\t$compass->{z}\t\n" ;

=head2 

=cut

    #print "\tACC_ANGLE: " . Dumper $accAngle ;


=head2 

    $minx = $accelerometer->{x} < $minx ? $accelerometer->{x} : $minx;
    $miny = $accelerometer->{y} < $miny ? $accelerometer->{y} : $miny;
    $minz = $accelerometer->{z} < $minz ? $accelerometer->{z} : $minz;
    $maxx = $accelerometer->{x} > $maxx ? $accelerometer->{x} : $maxx;
    $maxy = $accelerometer->{y} > $maxy ? $accelerometer->{y} : $maxy;
    $maxz = $accelerometer->{z} > $maxz ? $accelerometer->{z} : $maxz;
    #print "\tmax/min: $minx\t$miny\t$minz\t$maxx\t$maxy\t$maxz\n" ;
    my $acceleration_net = $accelerometer->{x}**2 + $accelerometer->{y}**2 + $accelerometer->{z}**2;
    my $net = ($acceleration_net**.5);
    $avg+=$net;
    $avgCount++;
    print "Net Acceleration : " . $net . "\n" ;
    print "Net Acceleration (avg): " . $avg/$avgCount . "\n" ;

=cut

    
    #print 'COMPASS: ' . Dumper {$dev->Compass->getRawReading()};
    #print 'ACCELEROMETER: ' . Dumper {$dev->Accelerometer->getRawReading()};
}



lib/Device/LSM303DLHC/Compass.pm
package Device::LSM303DLHC::Compass;

# PODNAME: Device::LSM303DLHC::Compass
# ABSTRACT: I2C interface to Compass on the LSM303DLHC 3 axis magnetometer(compass) and accelerometer using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use Moose;
use POSIX;

extends 'Device::SMBus';

=attr I2CDeviceAddress

Containd the I2CDevice Address for the bus on which your Compass is connected. It would look like 0x6b. Default is 0x1e.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x1e,
);

# Registers for the Magnetometer
use constant {
    MR_REG_M    => 0x02,
};

# X, Y and Z Axis magnetic Field Data value in 2's complement
use constant {
    OUT_X_H_M => 0x03,
    OUT_X_L_M => 0x04,

    OUT_Y_H_M => 0x07,
    OUT_Y_L_M => 0x08,

    OUT_Z_H_M => 0x05,
    OUT_Z_L_M => 0x06,
};

has magnetometerMaxVector => (
    is      => 'rw',
    default => sub {
        return (
            x => 0,
            y => 0,
            z => 0,);
    },
);

has magnetometerMinVector => (
    is      => 'rw',
    default => sub {
        return (
            x => 0,
            y => 0,
            z => 0,);
    },
);


=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(MR_REG_M,0x00);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from accelerometer registers

=cut

sub getRawReading {
    my ($self) = @_;

    return {
        x => $self->_typecast_int_to_int16( ($self->readByteData(OUT_X_H_M) << 8) | $self->readByteData(OUT_X_L_M) ),
        y => $self->_typecast_int_to_int16( ($self->readByteData(OUT_Y_H_M) << 8) | $self->readByteData(OUT_Y_L_M) ),
        z => $self->_typecast_int_to_int16( ($self->readByteData(OUT_Z_H_M) << 8) | $self->readByteData(OUT_Z_L_M) ),
    };
}

=method getReading

    $self->getReading()

Return proper calculated readings from the magnetometer

=cut

sub getReading {
    my ($self) = @_;
}

sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

1;


lib/Device/LSM303DLHC/Accelerometer.pm
package Device::LSM303DLHC::Accelerometer;

# PODNAME: Device::LSM303DLHC::Accelerometer
# ABSTRACT: I2C interface to Accelerometer on the LSM303DLHC 3 axis magnetometer(compass) and accelerometer using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use Moose;
use POSIX;

use Math::Trig;

extends 'Device::SMBus';

=attr I2CDeviceAddress

Containd the I2CDevice Address for the bus on which your Accelerometer is connected. It would look like 0x6b. Default is 0x19.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x19,
);

=attr gCorrectionFactor

This is a correction factor for converting raw values of acceleration in units of g or gravitational acceleration. It depends on the sensitivity set in the registers.

=cut

has 'gCorrectionFactor' => (
    is      => 'ro',
    default => 256
);

=attr gravitationAcceleration

This is the acceleration due to gravity in meters per second square usally represented as g. default on earth is around 9.8 although it differs from 9.832 near the poles to 9.780 at equator. This might also be different if you are on a different planet or in space.

=cut

has 'gravitationalAcceleration' => (
    is      => 'ro',
    default => 9.8
);

=attr mssCorrectionFactor

This attribute is built from the above two attributes automatically. This is usually gCorrectionFactor divided by gravitationalAcceleration. This is the inverse of relation between raw accelerometer values and its value in meters per seconds.

=cut

has 'mssCorrectionFactor' => (
    is         => 'ro',
    lazy_build => 1,
);

sub _build_mssCorrectionFactor {
    my ($self) = @_;
    $self->gCorrectionFactor/$self->gravitationalAcceleration;
}

use constant {
    PI => 3.14159265359,
};

# Registers for the Accelerometer 
use constant {
    CTRL_REG1_A => 0x20,
    CTRL_REG4_A => 0x23,
};

# X, Y and Z Axis magnetic Field Data value in 2's complement
use constant {
    OUT_X_H_A => 0x29,
    OUT_X_L_A => 0x28,

    OUT_Y_H_A => 0x2b,
    OUT_Y_L_A => 0x2a,

    OUT_Z_H_A => 0x2d,
    OUT_Z_L_A => 0x2c,
};


=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(CTRL_REG1_A,0b01000111);
    $self->writeByteData(CTRL_REG4_A,0b00101000);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from accelerometer registers

=cut

sub getRawReading {
    my ($self) = @_;

    use integer; # Use arithmetic right shift instead of unsigned binary right shift with >> 4
    my $retval = {
        x => ( $self->_typecast_int_to_int16( ($self->readByteData(OUT_X_H_A) << 8) | $self->readByteData(OUT_X_L_A) ) ) >> 4,
        y => ( $self->_typecast_int_to_int16( ($self->readByteData(OUT_Y_H_A) << 8) | $self->readByteData(OUT_Y_L_A) ) ) >> 4,
        z => ( $self->_typecast_int_to_int16( ($self->readByteData(OUT_Z_H_A) << 8) | $self->readByteData(OUT_Z_L_A) ) ) >> 4,
    };
    no integer;

    return $retval;
}

=method getAccelerationVectorInG

returns four acceleration vectors with accelerations in multiples of g - (9.8 meters per second square)
note that even when stationary on the surface of earth(or a earth like planet) there is a acceleration vector g that applies perpendicular to the surface of the earth pointing opposite of the surface. 

=cut

sub getAccelerationVectorInG {
    my ($self) = @_;
    
    my $raw = $self->getRawReading;
    return {
        x => ($raw->{x})/$self->gCorrectionFactor,
        y => ($raw->{y})/$self->gCorrectionFactor,
        z => ($raw->{z})/$self->gCorrectionFactor,
    };
}

=method getAccelerationVectorInMSS

returns four acceleration vectors with accelerations in meters per second square
note that even when stationary on the surface of earth(or a earth like planet) there is a acceleration vector g that applies perpendicular to the surface of the earth pointing opposite of the surface. 

=cut

sub getAccelerationVectorInMSS {
    my ($self) = @_;
    
    my $raw = $self->getRawReading;
    return {
        x => ($raw->{x})/$self->mssCorrectionFactor,
        y => ($raw->{y})/$self->mssCorrectionFactor,
        z => ($raw->{z})/$self->mssCorrectionFactor,
    };
}

=method getAccelerationVectorInclination

returns  coorinate angles between the acceleration vector(R) and the cartesian Coordinates(x,y,z). 

=cut

sub getAccelerationVectorAngles {
    my ($self) = @_;
    
    my $raw = $self->getRawReading;

    my $rawR = sqrt($raw->{x}**2+$raw->{y}**2+$raw->{z}**2); #Pythagoras theorem
    return {
        Axr => _acos($raw->{x}/$rawR),
        Ayr => _acos($raw->{y}/$rawR),
        Azr => _acos($raw->{z}/$rawR),
    };
}

=method getRollYawPitch

returns  Roll, Yaw and Pitch from the accelerometer. This is a bare reading from accelerometer and it assumes gravity is the only force on the accelerometer, which means it will be quiet inacccurate for a moving accelerometer.

=cut

sub getRollYawPitch {
    my ($self) = @_;
    
    my $raw = $self->getRawReading;

    return {
        Pitch => atan2($raw->{y},$raw->{z})+PI,
        Roll  => atan2($raw->{x},$raw->{z})+PI,
    };
}

sub _acos { 
    atan2( sqrt(1 - $_[0] * $_[0]), $_[0] ) 
}
sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

sub calibrate {
    my ($self) =@_;

}

1;


lib/Device/Magnetometer/LSM303DLHC.pm
use strict;
use warnings;
package Device::Magnetometer::LSM303DLHC;

# PODNAME: Device::Magnetometer::LSM303DLHC
# ABSTRACT: I2C interface to Magnetometer on the LSM303DLHC 3 axis magnetometer(compass) and accelerometer using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use POSIX;

use Moose;
extends 'Device::SMBus';

=attr I2CDeviceAddress

Contains the I2CDevice Address for the bus on which your Magnetometer is connected. It would look like 0x6b. Default is 0x1e.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x1e,
);

=register MR_REG_M

=cut

# Registers for the Magnetometer
use constant {
    MR_REG_M    => 0x02,
};

=register OUT_X_H_M

=register OUT_X_L_M

=register OUT_Y_H_M

=register OUT_Y_L_M

=register OUT_Z_H_M

=register OUT_Z_L_M

=cut

# X, Y and Z Axis magnetic Field Data value in 2's complement
use constant {
    OUT_X_H_M => 0x03,
    OUT_X_L_M => 0x04,

    OUT_Y_H_M => 0x07,
    OUT_Y_L_M => 0x08,

    OUT_Z_H_M => 0x05,
    OUT_Z_L_M => 0x06,
};

has magnetometerMaxVector => (
    is      => 'rw',
    default => sub {
        return {
            x => 424,
            y => 295,
            z => 472,
        };
    },
);

has magnetometerMinVector => (
    is      => 'rw',
    default => sub {
        return {
            x => -421,
            y => -639,
            z => -238,
        };
    },
);

=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(MR_REG_M,0x00);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from accelerometer registers

=cut

sub getRawReading {
    my ($self) = @_;

    return {
        x => $self->_typecast_int_to_int16( ($self->readByteData(OUT_X_H_M) << 8) | $self->readByteData(OUT_X_L_M) ),
        y => $self->_typecast_int_to_int16( ($self->readByteData(OUT_Y_H_M) << 8) | $self->readByteData(OUT_Y_L_M) ),
        z => $self->_typecast_int_to_int16( ($self->readByteData(OUT_Z_H_M) << 8) | $self->readByteData(OUT_Z_L_M) ),
    };
}

=method getMagnetometerScale1

    $self->getMagnetometerScale1()

Return proper calculated readings from the magnetometer scaled between +0.5 and
-0.5

=cut

sub getMagnetometerScale1 {
    my ($self) = @_;
    my $rawReading            = $self->getRawReading;
    my $magnetometerMaxVector = $self->magnetometerMaxVector;
    my $magnetometerMinVector = $self->magnetometerMinVector;
    return {
        x => ($rawReading->{x} - $magnetometerMinVector->{x})
            / ($magnetometerMaxVector->{x} - $magnetometerMinVector->{x}),
        y => ($rawReading->{y} - $magnetometerMinVector->{y})
            / ($magnetometerMaxVector->{y} - $magnetometerMinVector->{y}),
        z => ($rawReading->{z} - $magnetometerMinVector->{z})
            / ($magnetometerMaxVector->{z} - $magnetometerMinVector->{z}),
    };
}

sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

1;
